hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,311 @@
+use std::convert::TryFrom;
+
+use crate::{
+    error::{self},
+    indexed_merkle_proof::IndexedMerkleProof,
+    Digest,
+};
+
+use serde::Deserialize;
+
+use casper_types::{
+    allocate_buffer,
+    bytesrepr::{Bytes, FromBytes, ToBytes},
+};
+
+#[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProofDeserializeValidator {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl TryFrom<ChunkWithProofDeserializeValidator> for ChunkWithProof {
+    type Error = error::MerkleConstructionError;
+    fn try_from(value: ChunkWithProofDeserializeValidator) -> Result<Self, Self::Error> {
+        let candidate = Self {
+            proof: value.proof,
+            chunk: value.chunk,
+        };
+        if candidate.is_valid() {
+            Ok(candidate)
+        } else {
+            Err(error::MerkleConstructionError::IncorrectChunkProof)
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize)]
+#[schemars(with = "String", description = "Hex-encoded hash digest.")]
+#[serde(deny_unknown_fields, try_from = "ChunkWithProofDeserializeValidator")]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+
+        Ok((
+            ChunkWithProof::try_from(ChunkWithProofDeserializeValidator {
+                proof,
+                chunk: chunk.into(),
+            })
+            .map_err(|_| casper_types::bytesrepr::Error::Validation)?,
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    pub const CHUNK_SIZE: usize = 1;
+
+    // TODO[RC]: Uncomment when the `ChunkWithProof` is actually used in the code
+    // #[cfg(not(test))]
+    // pub const CHUNK_SIZE: usize = 1_048_576; // 2^20
```

Comment:
If you just re-export `ChunkWithProof`, you can uncomment this and get rid of the `#[cfg(test)]`